Replace `img` with `amp-pixel` when dealing with Facebook tracking pixel

delawski opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6059 

This PR changes the way the `AMP_Img_Sanitizer` works.

Whenever a Facebook tracking pixel is encountered, e.g.:

```html
<img height="1" width="1" style="display:none" alt="fbpx" src="https://www.facebook.com/tr?id=123456789012345&ev=PageView&noscript=1" />
```

it is is replaced with a more suitable `amp-pixel` instead of generic `amp-img`.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Replace `img` with `amp-pixel` when dealing with Facebook tracking pixel

westonruter commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -81,6 +81,7 @@ public function get_selector_conversion_mapping() {
 			Tag::IMG => [
 				'amp-img',
 				'amp-anim',
+				'amp-pixel',
```
This is a thoughtful addition, but in practice a Facebook pixel should really _never_ be styled by authors. It's supposed to be invisible after all. By adding `amp-pixel` to the list here, it means that if there is a `img { … }` it will get sanitized into `amp-img, amp-anim, amp-pixel { … }` .

```suggestion
```

westonruter commented on tests/php/test-amp-img-sanitizer.php:
```diff
@@ -48,13 +48,13 @@ public function test_get_selector_conversion_mapping() {
 		$with_defaults = new AMP_Img_Sanitizer( $dom );
 		$this->assertEquals(
-			[ 'img' => [ 'amp-img', 'amp-anim' ] ],
+			[ 'img' => [ 'amp-img', 'amp-anim', 'amp-pixel' ] ],
```
Per above:

```suggestion
			[ 'img' => [ 'amp-img', 'amp-anim' ] ],
```

westonruter commented on tests/php/test-amp-img-sanitizer.php:
```diff
@@ -48,13 +48,13 @@ public function test_get_selector_conversion_mapping() {
 		$with_false_native_used = new AMP_Img_Sanitizer( $dom, [ 'native_img_used' => false ] );
 		$this->assertEquals(
-			[ 'img' => [ 'amp-img', 'amp-anim' ] ],
+			[ 'img' => [ 'amp-img', 'amp-anim', 'amp-pixel' ] ],
```
Ditto:

```suggestion
			[ 'img' => [ 'amp-img', 'amp-anim' ] ],
```

westonruter commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -552,4 +567,19 @@ private function is_gif_url( $url ) {
+	 * @return bool Returns true if $url is a tracking pixel URL.
+	 */
+	private static function is_tracking_pixel_url( $url ) {
+		return (bool) preg_match( '/http(?>s)?:\/\/(?>www\.)?facebook\.com\/tr\?/i', $url );
```
Maybe parsing the URL would be a bit easier to work with?

```suggestion
		$parsed_url = wp_parse_url( $url );
		return isset( $parsed_url['host'], $parsed_url['path'] ) && 'facebook.com' === str_replace( 'www.',  $parsed_url['host'] ) && '/tr' === $parsed_url['path'];
```

westonruter commented on tests/php/test-amp-img-sanitizer.php:
```diff
@@ -923,4 +928,48 @@ public function test_process_picture_elements( $input, $args, $expected ) {
+	public function test_is_tracking_pixel_url( $url, $expected ) {
+		$this->assertEquals(
+			$expected,
+			$this->call_private_static_method( 'AMP_Img_Sanitizer', 'is_tracking_pixel_url', [ $url ] )
```
```suggestion
			$this->call_private_static_method( AMP_Img_Sanitizer::class, 'is_tracking_pixel_url', [ $url ] )
```

westonruter commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -183,6 +184,20 @@ public function sanitize() {
+			if ( self::is_tracking_pixel_url( $node->getAttribute( Attribute::SRC ) ) ) {
+				$amp_pixel_node = AMP_DOM_Utils::create_node(
+					$this->dom,
+					'amp-pixel',
```
You could add `use AmpProject\Extension;` to the top and then here do:

```suggestion
					Extension::PIXEL,
```

delawski replied:
Good idea. Plus, we already have `use AmpProject\Extension;` at the top. This has been addressed in 0d66f37.

delawski committed:
Remove `amp-pixel` from selector conversion mapping

In practice a Facebook pixel should really never be styled by authors. It's supposed to be invisible after all.

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Examine parsed URL instead of using regex

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Use `::class` constant instead of hardcoded class name

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Add missing replacement string to `str_replace` and reformat

delawski committed:
Use `Extension` class constant for getting `amp-pixel` tag name

delawski committed:
Remove redundant space

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/6059-img-to-amp-pixel

* 'develop' of github.com:ampproject/amp-wp: (63 commits)
  Improve nav menu E2E tests by creating test menu before each test suite
  Bail out if there is no menu location or it is already selected
  Add E2E tests for Twenty Twenty-Two header nav menu block
  Test Twenty Twenty search modal on mobile breakpoint
  Bump dependabot/fetch-metadata from 1.1.1 to 1.3.0
  Improve strings to account for one or more issues
  Fix type check by passing explicit string as href prop
  Include SCSS files in the lint-staged config for stylelint
  Fix stylelint issue
  Rename 'allowed' to 'associated'
  Use named functions
  Add label for why the validation data is being shown
  Remove wrapping of site-scan-results__source-detail
  Bump lint-staged from 12.3.4 to 12.3.5
  Update perimssions for gutenberg-packages-update workflow
  Add approve step for auto-merge workflow
  Add notice step
  Move conditional from matrix job to its steps
  Only run PHP feature tests when php files have changed
  Only run phpunit tests if a PHP file has changed
  ...

westonruter committed:
Copy referrerpolicy attribute from img to amp-pixel
